src/server/string_family.cc

```diff
@@ -693,76 +695,83 @@ void SetCmd::RecordJournal(const SetParams& params, string_view key, string_view
 }
 
 void StringFamily::Set(CmdArgList args, ConnectionContext* cntx) {
-  string_view key = ArgS(args, 0);
-  string_view value = ArgS(args, 1);
+  facade::CmdArgParser parser{args};
+
+  string_view key = parser.Next();
+  string_view value = parser.Next();
 
   SetCmd::SetParams sparams;
   sparams.memcache_flags = cntx->conn_state.memcache_flag;
 
-  int64_t int_arg;
-  SinkReplyBuilder* builder = cntx->reply_builder();
+  facade::SinkReplyBuilder* builder = cntx->reply_builder();
 
-  for (size_t i = 2; i < args.size(); ++i) {
-    ToUpper(&args[i]);
-
-    string_view cur_arg = ArgS(args, i);
+  while (parser.HasNext()) {
+    parser.ToUpper();
+    if (parser.Check("GET")) {
+      sparams.flags |= SetCmd::SET_GET;
+    } else if (parser.Check("STICK")) {
+      sparams.flags |= SetCmd::SET_STICK;
+    } else if (parser.Check("KEEPTTL")) {
+      sparams.flags |= SetCmd::SET_KEEP_EXPIRE;
+    } else if (parser.Check("XX")) {
+      sparams.flags |= SetCmd::SET_IF_EXISTS;
+    } else if (parser.Check("NX")) {
+      sparams.flags |= SetCmd::SET_IF_NOTEXIST;
+    } else {
+      string_view opt = parser.Next();
+      if (opt == "EX" || opt == "PX" || opt == "EXAT" || opt == "PXAT") {
+        int64_t int_arg = parser.Next<int64_t>();
+        if (auto err = parser.Error(); err) {
+          return builder->SendError(err->MakeReply());
+        }
```

Comment:
Also if you're just experimenting with the parser, the following is possible

if (parser.Peek() in {"EX", "PX", ... }) {
  // handle expiry
} else { // must be flag
  uint16_t flag = parser.Switch("GET", SetCmd::SET_GET, "STICK", SetCmd::SET_STICK....);
  flags |= flag; // returns 0 if it failed to parse, checked with Error at the end
}

if (auto err = parser.Error(); err) 
  return cntx->SendError(err->MakeReply())

Reply:
nice. changed